File: src/CXX/abstract_load_balancer.cxx
#include <SLB4MPI/abstract_load_balancer.hpp>

#include <cstdlib>
#include <algorithm>

/**
 *
 * @brief initialize default values of abstract load balancer
 *
 * @param[in]           communicator   - MPI communicator on which load balancer will be used
 * @param[in]           lower_bound    - lower bound of range
 * @param[in]           upper_bound    - upper bound of range, upper_bound >= lower_bound
 * @param[in, optional] min_chunk_size - minimal size of chank that can be associated with job, default: 1
 * @param[in, optional] max_chunk_size - maximal size of chank that can be associated with job, default: upper_bound - lower_bound + 1
 *
 */
SLB4MPI::AbstractLoadBalancer::AbstractLoadBalancer(const int communicator, const int64_t lower_bound, const int64_t upper_bound, const int64_t min_chunk_size, const int64_t max_chunk_size) {
  this->communicator = communicator;
  this->rank = 0;
  this->nranks = 1;
  this->root = 0;

#ifdef SLB4MPI_WITH_MPI
  MPI_Comm_rank(this->communicator, &(this->rank));
  MPI_Comm_size(this->communicator, &(this->nranks));
#endif

  this->lower_bound = lower_bound;
  this->upper_bound = upper_bound;

  this->min_chunk_size = min_chunk_size;
  this->max_chunk_size = max_chunk_size;
  if (this->max_chunk_size < 0) {
    this->max_chunk_size = std::max((this->upper_bound - this->lower_bound + 1) / this->nranks, this->min_chunk_size);
  }

  if (this->min_chunk_size <= 0) std::exit(1); // "Min chunk size must be greater than zero!"

  if (this->lower_bound > this->upper_bound) std::exit(1); // "Upper bound is less than lower bound!"
  if (this->min_chunk_size > this->max_chunk_size) std::exit(1); // "Max chunk size is less than min chunk size!"
}

File: src/Fortran/abstract_load_balancer.f90
module SLB4MPI_abstract_load_balancer_m
#ifdef SLB4MPI_WITH_MPI
  use mpi
#endif
  implicit none
  private

#ifndef SLB4MPI_WITH_MPI
  integer, parameter :: MPI_INTEGER_KIND = 4
  integer, parameter :: MPI_ADDRESS_KIND = 8
#endif
  public :: MPI_INTEGER_KIND
  public :: MPI_ADDRESS_KIND
  integer, parameter, public :: BITS_IN_BYTE = 8

  type, abstract :: load_balancer_t
    integer(MPI_INTEGER_KIND) :: communicator !< MPI communicator
    integer(MPI_INTEGER_KIND) :: rank         !< Rank of the process
    integer(MPI_INTEGER_KIND) :: root = 0     !< Rank of the root process
    integer(MPI_INTEGER_KIND) :: nranks       !< Number of processes (group size)
    integer(8) :: lower_bound                 !< Lower bound of range
    integer(8) :: upper_bound                 !< Upper bound of range
    integer(8) :: min_chunk_size              !< Minimal chunk size for job
    integer(8) :: max_chunk_size              !< Maximal chunk size for job
  contains
    procedure(initialize), deferred, public :: initialize
    procedure(get_range),  deferred, public :: get_range
    procedure(clean),      deferred, public :: clean

    procedure :: default_initialize
  end type

  abstract interface

    !>
    !> @brief initialize load balancer
    !>
    !> @param[in,out] lb         - load balancer object
    !> @param[in] communicator   - MPI communicator on which load balancer will be used
    !> @param[in] lower_bound    - lower bound of range
    !> @param[in] upper_bound    - upper bound of range, upper_bound >= lower_bound
    !> @param[in] min_chunk_size - minimal size of chank that can be associated with job, default: 1
    !> @param[in] max_chunk_size - maximal size of chank that can be associated with job, default: upper_bound - lower_bound + 1
    !>
    subroutine initialize(lb, communicator, lower_bound, upper_bound, min_chunk_size, max_chunk_size)
      import load_balancer_t, MPI_INTEGER_KIND
      class(load_balancer_t),    intent(inout) :: lb
      integer(MPI_INTEGER_KIND), intent(in)    :: communicator
      integer(8),                intent(in)    :: lower_bound, upper_bound
      integer(8),      optional, intent(in)    :: min_chunk_size, max_chunk_size
    end subroutine initialize

    !>
    !> @brief get range to compute
    !>
    !> @note `upper_bound` can be only less than or equal `load_balancer_t%upper_bound`.
    !>       So, for last elements a batch with size less than `min_chunk_size` can be returned
    !>
    !> @param[in,out] lb       - load balancer object
    !> @param[out] lower_bound - lower bound of range to compute
    !> @param[out] upper_bound - upper bound of range to compute
    !> @return                 - true if there is something to compute
    !>
    logical function get_range(lb, lower_bound, upper_bound)
      import load_balancer_t
      class(load_balancer_t), intent(inout) :: lb
      integer(8),             intent(out)   :: lower_bound, upper_bound
    end function get_range

    !>
    !> @brief load balancer destructor
    !>
    !> @param[in,out] lb - load balancer object
    !>
    subroutine clean(lb)
      import load_balancer_t
      class(load_balancer_t), intent(inout) :: lb
    end subroutine clean

  end interface

  public :: load_balancer_t

contains

  !>
  !> @brief initialize variables of load balancer with default values
  !>
  !> @param[in,out] lb         - load balancer object
  !> @param[in] communicator   - MPI communicator on which load balancer will be used
  !> @param[in] lower_bound    - lower bound of range
  !> @param[in] upper_bound    - upper bound of range, upper_bound >= lower_bound
  !> @param[in] min_chunk_size - minimal size of chank that can be associated with job, default: 1
  !> @param[in] max_chunk_size - maximal size of chank that can be associated with job, default: upper_bound - lower_bound + 1
  !>
  subroutine default_initialize(lb, communicator, lower_bound, upper_bound, min_chunk_size, max_chunk_size)
    class(load_balancer_t),    intent(inout) :: lb
    integer(MPI_INTEGER_KIND), intent(in)    :: communicator
    integer(8),                intent(in)    :: lower_bound, upper_bound
    integer(8),      optional, intent(in)    :: min_chunk_size, max_chunk_size

    integer(MPI_INTEGER_KIND) :: ierr

#ifdef SLB4MPI_WITH_MPI
    call MPI_Comm_rank(communicator, lb%rank, ierr)
    call MPI_Comm_size(communicator, lb%nranks, ierr)
#else
    lb%rank = 0
    lb%nranks = 1
#endif

    lb%communicator = communicator
    lb%lower_bound = lower_bound
    lb%upper_bound = upper_bound
    lb%min_chunk_size = 1
    if (present(min_chunk_size)) lb%min_chunk_size = min_chunk_size
    lb%max_chunk_size = max((lb%upper_bound - lb%lower_bound + 1) / lb%nranks, lb%min_chunk_size)
    if (present(max_chunk_size)) lb%max_chunk_size = max_chunk_size

    if (lb%min_chunk_size <= 0) error stop "Min chunk size must be greater than zero!"

    if (lb%lower_bound > lb%upper_bound) error stop "Upper bound is less than lower bound!"
    if (lb%min_chunk_size > lb%max_chunk_size) error stop "Max chunk size is less than min chunk size!"

  end subroutine default_initialize
end module SLB4MPI_abstract_load_balancer_m
